### PR TITLE
keynames: fix combining dot-dot parts and escapes

### LINKED ISSUE #3574: keyname overhaul: crash
```sh
kdb export system:/elektra/mountpoints toml
```

leads to following backtrace:

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: Datei oder Verzeichnis nicht gefunden.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007efc5133a535 in __GI_abort () at abort.c:79
#2  0x0000562fe7e14f6c in catchSignal (signum=<optimized out>) at ./src/tools/kdb/main.cpp:110
#3  <signal handler called>
#4  elektraKeyNameCanonicalize (name=0x7efc507e64ba "", name@entry=0x7efc507e64b8 "..", canonicalName=canonicalName@entry=0x562fe81760e0, canonicalSizePtr=canonicalSizePtr@entry=0x562fe81760e8, offset=9, usizePtr=usizePtr@entry=0x562fe81760f8)
    at ./src/libs/elektra/keyname.c:844
#5  0x00007efc516815b9 in keyAddName (key=key@entry=0x562fe81760d0, newName=newName@entry=0x7efc507e64b8 "..") at ./src/libs/elektra/keyname.c:565
#6  0x00007efc507dd16d in addMissingArrayKeys (parent=0x562fe8233990, keys=0x562fe8232310) at ./src/plugins/toml/prepare.c:147
#7  prepareKeySet (keys=keys@entry=0x562fe8232310, parent=parent@entry=0x562fe8233990) at ./src/plugins/toml/prepare.c:39
#8  0x00007efc507dc877 in tomlWrite (keys=0x562fe8232310, parent=0x562fe8233990) at ./src/plugins/toml/write.c:82
#9  0x00007efc507d85ff in elektraTomlSet (handle=<optimized out>, returned=<optimized out>, parentKey=<optimized out>) at ./src/plugins/toml/toml.c:49
#10 0x00007efc51703e1c in kdb::tools::Plugin::set (this=<optimized out>, ks=..., parentKey=...) at ./src/bindings/cpp/include/keyset.hpp:596
#11 0x0000562fe7dbfdc1 in ExportCommand::execute (this=<optimized out>, cl=...) at /usr/include/c++/8/bits/unique_ptr.h:342
#12 0x0000562fe7da20bd in main (argc=<optimized out>, argv=0x7ffceef7e5d8) at /usr/include/c++/8/bits/unique_ptr.h:342
```

relevant bt full:

```
#4  elektraKeyNameCanonicalize (name=0x7efc507e64ba "", name@entry=0x7efc507e64b8 "..", canonicalName=canonicalName@entry=0x562fe81760e0, canonicalSizePtr=canonicalSizePtr@entry=0x562fe81760e8, offset=9, usizePtr=usizePtr@entry=0x562fe81760f8)
    at ./src/libs/elektra/keyname.c:844
        i = 1
        backslashes = <optimized out>
        newOutPtr = 0x0
        ulen = 94763757494792
        end = <optimized out>
        len = <optimized out>
        nameLen = <optimized out>
        outPtr = 0x1 <error: Cannot access memory at address 0x1>
        usize = 1
        rootOffset = <optimized out>
#5  0x00007efc516815b9 in keyAddName (key=key@entry=0x562fe81760d0, newName=newName@entry=0x7efc507e64b8 "..") at ./src/libs/elektra/keyname.c:565
No locals.
#6  0x00007efc507dd16d in addMissingArrayKeys (parent=0x562fe8233990, keys=0x562fe8232310) at ./src/plugins/toml/prepare.c:147
        name = 0x562fe81760d0
        arrays = 0x562fe81a0540
        key = <optimized out>
        arrays = <optimized out>
        key = <optimized out>
        name = <optimized out>
        index = <optimized out>
        arrayRoot = <optimized out>
        next = <optimized out>
```

@kodebach does this already help?

### PR DESCRIPTION
**Pull request by kodebach**
closes #3574

## Basics

These points need to be fulfilled for every PR:

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy.
- [x] The PR is rebased with current master.

## Checklist

- [x] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation
- [x] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [x] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [x] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

### CODE REVIEW

**Comment by kodebach** [comment]
@mpranj Since this is a somewhat critical issue (see #3574), this should probably be merged ASAP.

File: doc/news/_preparation_next_release.md [hunk 1]
```diff
@@ -43,6 +43,8 @@ If you rely on specific behaviour of Elektra's Keynames and have already taken t
 
 <!-- TODO (kodebach): finish -->
 
+. _(Klemens Böswirth)_
```
Comment (mpranj) [inline doc/news/_preparation_next_release.md:46]: 🥇

**Comment by mpranj** [comment]
> @mpranj Since this is a somewhat critical issue (see #3574), this should probably be merged ASAP.

Ok, but let's first wait for a few builds to finish.

**Comment by mpranj** [comment]
Thank you for fixing it so quickly!

**Comment by kodebach** [comment]
> Thank you for fixing it so quickly!

Well it was quite an obvious problem, once I added the relevant test cases in `test_keyname.c` and did a debug run :sweat_smile:

**Pull request merged**